keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,692 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, `kube-apiserver` allows one to configure the authorization chain using
+a set of command line flags of the format `--authorization-*`. And cluster
+administrators can only enable one webhook as a part of the authorization
+chain using the `--authorization-modes` flag. This limits admins from creating
+authorization chains with multiple webhooks that validate requests in a certain
+order. This proposal makes the case for a more structured config to define the
+authorization chain for `kube-apiserver` and allowing multiple webhooks with
+well-defined parameters and enabling fine grained control like explicit Deny
+authorizer.
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization
+webhook in the API Server handler chain. They also need to be able to
+declaratively configure the authorizer chain using a configuration file. It
+should also be easy to say when to Deny requests, for example, when a webhook
+is unreachable.
+
+### Goals
+
+- Define a configuration file format for configuring Kubernetes API Server
+Authorization chain.
+- Allow ordered definition of authorization modes.
+- Allow definition of multiple webhooks in the authorization chain.
+- Allow resource/user based pre-filtering of webhooks using CEL to prevent unnecessary
+invocations.
+- Enable user to define the policy when a webhook can't be reached due to
+network errors etc.
+- Enable hot reload of the configuration.
+
+### Non-Goals
+
+- Replace path to kubeconfig file for webhooks by allowing auth configuration
+within the new file structure
+- Allow ordering of super user authorizer (always allow user group;
+`system:masters`) so that it can be moved from the beginning of the auth chain.
+- Deciding whether ABAC should be continued to be used by the users.
+
+> Note: The above may be revisited in future alpha iterations of this KEP or as a separate KEP (for e.g., the superuser authorizer)
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain. See [Design Details](#design-details) for more details.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+  - type: Webhook
+    webhook:
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      failurePolicy: Deny
+      connectionInfo:
+        type: KubeConfig
+        kubeConfigFile: /kube-system-authz-webhook.yaml
+      matchConditions:
+      - expression: |
+        has(request.resourceAttributes) && (request.resourceAttributes.namespace == 'kube-system')
+      - expression: |
+        !('system:serviceaccounts:kube-system' in request.user.groups)
+  - type: Node
+  - type: RBAC
+  - type: Webhook
+    webhook:
+      authorizedTTL: 5m
+      unauthorizedTTL: 30s
+      timeout: 3s
```

Comment:
I had a concern that's got lost - https://github.com/kubernetes/enhancements/pull/3376#discussion_r1204487517

I don't have any concern about this going to alpha. Could we track it as a detail to make sure we consider before this moves to GA?

Reply:
there are only two examples of using field names with units in file-based config

```
NodeLeaseDurationSeconds int32 `json:"nodeLeaseDurationSeconds,omitempty"`
ShutdownGracePeriodSeconds int64 `json:"shutdownGracePeriodSeconds"`
PodInitialBackoffSeconds *int64 `json:"podInitialBackoffSeconds,omitempty"`
PodMaxBackoffSeconds *int64 `json:"podMaxBackoffSeconds,omitempty"`
```

there are 35+ examples of using metav1.Duration in config files, including for things like kubelet webhook authentication:
```
CacheTTL metav1.Duration `json:"cacheTTL,omitempty"`
```
credentialprovider cache duration:
```
CacheDuration *metav1.Duration `json:"cacheDuration,omitempty"`
```

I think the time.Duration type is more usable and more consistent. I don't think we'll change this.

Reply:
> validation around the timeout so that it cannot exceed 30 seconds

Ah, nice.